### DISCUSSION TITLE
Best practice for Typescript + Slice Pattern + Immer + Devtools

### DISCUSSION THREAD

**Discussion by utkubiz** [pmndrs/zustand#2070] [Q&A] [answered]
Hello everyone,

I've set up a structure as follows. In this structure, I'm using both the slice pattern and immer + devtools middleware. 
Everything is currently working smoothly. However, what are your thoughts on this structure? 

Is this the best practice for this situation, or do you know of a better solution?

https://tsplay.dev/weLJeW

**Comment by dbritto-dev** [comment, marked as answer]
@utkubiz looks good to me. I would do the same. You are on the right path. +1 for `ImmerStateCreator`

**Comment by Itisfilipe** [comment]
Great! This was exactly what I was looking for. Thank you!

**Comment by vorant94** [comment]
nice, looks good!

the only thing i cannot understand is how applying devtools on the resulting store is getting into autocomplete of related props on the slice level... how does TS know that there can be a third argument for set function?

```ts
export const practiceSlice: ImmerStateCreator<PracticeState> = (set) => ({
	activeQuestion: 0,
	nextQuestion: () =>
		set(
			(state) => {
				state.practice.activeQuestion++;
			},
			undefined,
			{ type: "nextQuestion" },
		),
});
```

Is it coming from `StoreCreator`? If so, how does it understand when it will be used with or without devtools?
